File: payme/python/payme_payment/payme_checkout.py
"""Payme to'lov havolasini (checkout link) qurish.

Click'dan farqli — Payme havolasi query-string emas, parametrlarni
`key=value;key=value` shaklida yig'ib, BUTUN satrni base64 qiladi:

    https://checkout.paycom.uz/<base64("m=MERCHANT_ID;ac.order_id=42;a=500000")>

Bu faylga tegishingiz shart emas.
"""

from __future__ import annotations

import base64
from decimal import Decimal

from .payme_config import PaymeConfig, get_config


def som_to_tiyin(som: Decimal | int | float | str) -> int:
    """So'mni tiyinga o'giradi (1 so'm = 100 tiyin). Payme SUMMA TIYINDA kutadi."""
    return int(round(Decimal(str(som)) * 100))


def checkout_url(
    account: dict[str, str | int],
    amount_tiyin: int,
    *,
    return_url: str | None = None,
    lang: str | None = None,
    config: PaymeConfig | None = None,
) -> str:
    """Foydalanuvchi yuboriladigan Payme to'lov havolasini quradi.

    `account` — Payme'ga yuboriladigan hisob maydonlari (masalan
    `{"order_id": "42"}`). Shu maydonlar `payme_orders.py`dagi
    `find_account()` ga qaytib keladi — ular orqali buyurtmani topasiz.

    `amount_tiyin` — summa TIYINDA (so'm emas!). `som_to_tiyin()` bilan
    o'giring: `checkout_url({"order_id": 42}, som_to_tiyin(5000))`.

    Namuna:
        url = checkout_url({"order_id": 42}, som_to_tiyin(5000))
        # foydalanuvchini shu havolaga yuboring
    """
    cfg = config or get_config()

    parts = [f"m={cfg.merchant_id}"]
    for key, value in account.items():
        parts.append(f"ac.{key}={value}")
    parts.append(f"a={int(amount_tiyin)}")

    if return_url:
        parts.append(f"c={return_url}")
    if lang:
        parts.append(f"l={lang}")

    raw = ";".join(parts)
    encoded = base64.b64encode(raw.encode("utf-8")).decode("ascii")

    return f"{cfg.checkout_base_url}/{encoded}"

File: payme/php/payme_checkout.php
<?php
/**
 * Payme to'lov havolasini (checkout link) qurish.
 *
 * Click'dan farqli — Payme havolasi query-string emas, parametrlarni
 * `key=value;key=value` shaklida yig'ib, BUTUN satrni base64 qiladi:
 *
 *     https://checkout.paycom.uz/<base64("m=MERCHANT_ID;ac.order_id=42;a=500000")>
 *
 * Bu faylga tegishingiz shart emas.
 */

require_once __DIR__ . '/payme_config.php';

if (!function_exists('paymeSomToTiyin')) {
    /** So'mni tiyinga o'giradi (1 so'm = 100 tiyin). Payme SUMMA TIYINDA kutadi. */
    function paymeSomToTiyin($som)
    {
        return (int)round(((float)$som) * 100);
    }
}

if (!function_exists('paymeCheckoutUrl')) {
    /**
     * Foydalanuvchi yuboriladigan Payme to'lov havolasini quradi.
     *
     * `$account` — Payme'ga yuboriladigan hisob maydonlari (masalan
     * `['order_id' => 42]`). Shu maydonlar `payme_orders.php`dagi
     * `paymeFindAccount()` ga qaytib keladi.
     *
     * `$amountTiyin` — summa TIYINDA (so'm emas!). `paymeSomToTiyin()`
     * bilan o'giring.
     *
     * Namuna:
     *     $url = paymeCheckoutUrl(['order_id' => 42], paymeSomToTiyin(5000));
     */
    function paymeCheckoutUrl(array $account, $amountTiyin, $returnUrl = null, $lang = null)
    {
        $parts = array('m=' . paymeConfig('merchant_id'));

        foreach ($account as $key => $value) {
            $parts[] = 'ac.' . $key . '=' . $value;
        }

        $parts[] = 'a=' . (int)$amountTiyin;

        if ($returnUrl !== null) {
            $parts[] = 'c=' . $returnUrl;
        }
        if ($lang !== null) {
            $parts[] = 'l=' . $lang;
        }

        $raw = implode(';', $parts);
        $encoded = base64_encode($raw);

        return rtrim(paymeConfig('checkout_base_url'), '/') . '/' . $encoded;
    }
}
